Addition of schur number in combinatorics

mijo2 opened the pull request:
#### References to other Issues or PRs
Revives #14493 
Closes #14493 

#### Brief description of what is fixed or changed
--> Added schur number in combinatorics which computes and returns the sum-free subsets based on the lower bound of the schur number

#### Other comments
For example:
```
>>> from sympy import *
>>> from sympy.abc import *
>>> from sympy.combinatorics.schur_number import *
>>> schur_partition(2)
[[1, 2]]
>>> schur_partition(5)
[[3, 2], [5], [1, 4]]
```

#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* combinatorics
  * Added schur number utilities in combinatorics
<!-- END RELEASE NOTES -->

slacker404 committed:
added feature Schur_number

slacker404 committed:
misc typos

slacker404 committed:
Added symbolic function for Schur_lower_bound and modified test_schur_number

mijo2 committed:
Merge branch 'master' of https://github.com/slacker404/sympy into 14493_schur_number

mijo2 commented:
Things to be done:
1. Update the schur number to have lower computations as there maybe some redundant ones.
2. Address the suggestions by @asmeurer 
3. Add more than one test cases.

czgdp1807 commented on sympy/combinatorics/tests/test_schur_number.py:
```diff
@@ -0,0 +1,29 @@
+                raise AssertionError("This subset is not sum free")
+
+def test_schur_number():
+    n = random.randint(1,1000)
```
What about setting the seed for deterministic tests?

mijo2 replied:
@czgdp1807 I was planning to do that soon. Adding some deterministic and some random tests.

czgdp1807 replied:
IMO, adding non-deterministic tests would put the master build at risk. I would say set the set if it isn't already set for any test using random.

mijo2 replied:
Ok @czgdp1807 I will set the seed values and probably run these random tests 4 to 5 times.

mijo2 committed:
combinatorics: Removed unncessary computations

Removed unnecessay computations in schur_number.py

mijo2 committed:
combinatorics: Added seed value for using random

Added seed value for using random in test_schur_number.py and
increased the number of tests to 5.

czgdp1807 commented on sympy/combinatorics/tests/test_schur_number.py:
```diff
@@ -0,0 +1,32 @@
+            t += len(item)
+            for l in item:
+                if l in numbers:
+                    raise AssertionError("Every number must exist once")
```
`assert l in numbers`

mijo2 replied:
@czgdp1807 I think it would be ```assert l in numbers is False``` since if l is in numbers, we are raising an assertion error.

czgdp1807 commented on sympy/combinatorics/tests/test_schur_number.py:
```diff
@@ -0,0 +1,32 @@
+    for i in subset:
+        for j in subset:
+            if i + j in subset:
+                raise AssertionError("This subset is not sum free")
```
`assert i + j in subset is False`

czgdp1807 commented on sympy/combinatorics/schur_number.py:
```diff
@@ -0,0 +1,73 @@
+    in order to find the Schur Number
+    """
+    n = Symbol("n")
+    k = Symbol("k")
```
`n, k = symbols("n, k")`.
Are they integers?

mijo2 replied:
Yes, ```n``` and ```k``` are supposed to be integers since this formula works only with integers.

czgdp1807 replied:
Then what about, `n, k = symbols("n, k", integer=True)`?

mijo2 replied:
Ok yes I will update this, they are integers anyways and it makes sense for defining them as such.

czgdp1807 commented on sympy/combinatorics/schur_number.py:
```diff
@@ -0,0 +1,73 @@
+        temp_2 = [number*3 - 1 for number in item if number*3 - 1 <= n]
+        new_item = temp_1 + temp_2
+        new_list.append(new_item)
+    last_list = [3*k + 1 for k in range(0, len(current_list)+1) if 3*k +1 <= n]
```
Split into two lines.

czgdp1807 commented on sympy/combinatorics/schur_number.py:
```diff
@@ -0,0 +1,73 @@
+    subsets since the only known Schur numbers are:
+    S(1) = 1, S(2) = 4 , S(3) = 13, S(4) = 44.
+    e.g for n = 44 the lower bound from the function above is 5 subsets but it has been proven
+    that can be done with 4 subsets.
```
IMO, examples should be added.

czgdp1807 commented on sympy/combinatorics/schur_number.py:
```diff
@@ -0,0 +1,73 @@
+def schur_number_lower_bound(n):
+    """
+    This function returns a lower bound to Schur's Number
+    """
```
+1

mijo2 replied:
Ok @czgdp1807 I will make these changes soon.

mijo2 committed:
combinatorics: Docstring and some minor updates

Docstring of this component was updated to have more examples along
with some minor updates and changes in test_schur_number.py

czgdp1807 commented:
LGTM. Just address the remaining comments.

mijo2 committed:
combinatorics: Made the schur lower bound number into function

mijo2 commented:
@czgdp1807 @asmeurer I have came up with a function and have tried to address [this comment](https://github.com/sympy/sympy/pull/14493#issuecomment-376251946). Please take a look if possible.

asmeurer commented on sympy/combinatorics/tests/test_schur_number.py:
```diff
@@ -0,0 +1,30 @@
+    random.seed(1000)
+
+    for _ in range(5):
+        n = random.randint(1, 1000)
```
This should use sympy.utilities.randtest

asmeurer commented on sympy/combinatorics/schur_number.py:
```diff
@@ -0,0 +1,120 @@
+
+    return current_list
+
+def Schur_number_lower_bound():
```
What is the point of this function?

mijo2 replied:
Ohh I have to remove it, I will remove this in the next PR.

asmeurer replied:
You don't need to make another PR. Just push up commits to this same branch and this PR will update.

mijo2 replied:
@asmeurer I am sorry I meant commit

asmeurer commented on sympy/combinatorics/schur_number.py:
```diff
@@ -0,0 +1,120 @@
+
+class schur_number_lower_bound(Function):
+    """
+    Gives the value for 1/2 * (3**n  - 1)
```
I don't see the point of this either. This is such a simple expression you could just use it directly wherever it is needed (it looks like it isn't actually used elsewhere in this module).

mijo2 replied:
It is for the user to get the lower bound of ```S(n)``` i.e the schur number.

asmeurer commented on sympy/combinatorics/schur_number.py:
```diff
@@ -0,0 +1,120 @@
+
+    @classmethod
+    def eval(cls, k):
+        expr = (1/2) * (3**k - 1)
```
Don't write 1/2 directly in a SymPy expression. That will cause it to create a float. See https://docs.sympy.org/latest/tutorial/gotchas.html. A better way to write this is `(3**k - 1)/2`.

mijo2 replied:
@asmeurer I will update this in the next PR.

asmeurer commented on sympy/combinatorics/schur_number.py:
```diff
@@ -0,0 +1,120 @@
+
+def schur_partition(n):
+    """
+    This function makes the partition in the minimum number of sum-free subsets
```
This docstring needs to follow the formatting outlined in the style guide https://docs.sympy.org/latest/documentation-style-guide.html

asmeurer replied:
This description does not make it clear at all what this function returns. Does it return subsets, or a number. If subsets, what form are they?

mijo2 replied:
It returns the subsets in a list of lists format and it is mentioned in the second sentence of the explanation. I will add new lines between the first and the second sentences to make it look clear.

asmeurer commented:
The functions here would be better as generators.

asmeurer commented:
Does this fit into an existing file (I'm not sure if it does, as I'm not too familiar with the combinatorics module)?

mijo2 commented:
@asmeurer I don't know have enough experience with ```combinatorics``` module either but having a glance of the files in the module, I think it probably requires its own file.

mijo2 commented:
@asmeurer What are generators exactly? I mean are you referring to generators in ```combinatorics``` module?

asmeurer commented:
Generators use `yield` to return each individual element one at a time, rather than the whole list. Here is some more details https://stackoverflow.com/questions/231767/what-does-the-yield-keyword-do. If you look in the combinatorics module or in sympy.utilities.iterables you will see lots of examples.

mijo2 commented:
@asmeurer Thanks for the information about generators, this helped a lot. But I don't think that generators would be much helpful here for both the functions (```schur_partition``` and ```_generate_next_list```) as ```_generate_next_list``` has to return the list as if it yields then anyway,  we will have to form a list in ```schur_partition``` as for the line ```sum_free_subsets[-1] += missed_elements``` and as for ```schur_partition```, it is not sure until the computation ends as to whether an element that we want to yield will be one of the final sum-free subsets or not.

mijo2 committed:
combinatorics: Some suggested changes were incorporated

Docstring for schur_partition was updated, test_args test was added
for the two new subclasses of Function introduced in this PR,
redundant functions were removed.

mijo2 committed:
combinatorics: Minor bug fix

Unused import was removed. This commit will be squashed later.

sylee957 commented:
It's weird to see classes named so long like `schur_number_lower_bound`.
I think that the true Schur number is sort of a conjecture problem. But I agree that it's not worth creating an another weird function just to model some criterion.

So I would rather think that it should be made into a formal symbol like `class SchurNumber(Function)`, 
And be designed like
```
SchurNumber(1) => 2
SchurNumber(2) => 5
...
SchurNumber(5) => 161
SchurNumber(6) => SchurNumber(6)
...
```

And `schur_number_lower_bound` should rather be wrapped to the inequality computation of SymPy like
`SchurNumber(n) > (3**n - 1)/2 => True`

mijo2 commented:
@sylee957 I have two questions regarding this suggestion:
  1. By wrapping the inequality computation of SymPy, do you mean something like below:
      ```
      def schur_lower_bound(k):
           return SchurNumber(k) > (3**k - 1)/2
      ```
     Or something like actually wrapping it up with decorators.
  2. What action should we take for ```schur_number_subsets_lower_bound```?

sylee957 commented:
You may have to modify `SchurNumber.__gt__` such that it can check RHS is smaller than `(3**k - 1)/2`.
It would also be good to put `lower_bound_schur` as a method under `SchurNumber` class I design.
I just don't want to see `(3**k - 1)/2` added into an another trivial function class.

`schur_number_subsets_lower_bound` can also be implemented as a double check under `__gt__`. But I just want to confirm which source is it from.

sylee957 commented:
But first, I would want to see the `SchurNumber` class to be designed as I suggest (To give only some few known values, and not evaluating otherwise).
The advantage of symbolic math, is that you should be able to design things that you cannot even compute.

mijo2 commented:
@sylee957 source for ```Schur number``` is [this](http://mathworld.wolfram.com/SchurNumber.html). So, I am planning to build something like below - 
```
class SchurNumber(Function):

    @classmethod
    def eval(cls, k):
        if k.is_Number:
            if k is S.Infinity:
                return S.Infinity
            if k.is_zero:
                return 0
            if not k.is_Integer or k.is_negative:
                raise ValueError("k should be a positive integer")
            first_known_schur_numbers = {1: 1, 2: 4, 3: 13, 4: 44}
            if k <= 4:
                return Integer(first_known_schur_numbers[k])

    def lower_bound(self):
        if len(self.args) > 1 or (iterable(self.args[0]) and len(self.args[0]) > 1):
            raise ValueError("Multiple arguments not allowed.")
        f_ = self.args[0]
        return (3**f_ - 1)/2
```

sylee957 commented:
Yes, that is the way I’ve suggested.

I don’t think that you would have to check the length of args when computing the lower_bound, since functions already sanitizes such so.

I think that k.is_Integer should be relaxed to k.is_integer to allow symbolic integers.
And then it’s possible to query stuff like lower_bounds to give symbolic result.

mijo2 commented:
> I don’t think that you would have to check the length of args when computing the lower_bound, since functions already sanitizes such so.

Ok, I will remove the check.

> I think that k.is_Integer should be relaxed to k.is_integer to allow symbolic integers.

Ok will relax this.

mijo2 commented:
@sylee957 I will be a little inactive for three days and this is a big update to fix. I will continue this PR after my exams.

mijo2 commented:
@sylee957 what about the condition ```.is_Number```. Should it be ```.is_Number```?

sylee957 commented:
I don’t think that it is needed for most of the checks, because the comparison with the infinity should be singleton comparison, and other cases can include symbolics.

It may only be needed before k <= 4

mijo2 committed:
combinatorics: Added updates for class definitions and other changes

Updated the class definitions and names along with the entire
transformation of SchurNumber function. Updated the schur_partition
to raise a ValueError when the input symbol is a variable.

mijo2 commented:
@sylee957 I have updated the classes as you suggested. You can take a look. Also, I had a question for ```_generate_next_list``` function. Now the function's whole list that it returns is ready only till the last iteration, so is it a good idea to make it a generator function?

mijo2 committed:
combinatorics: Minor update for rebuild

sylee957 commented on sympy/combinatorics/schur_number.py:
```diff
@@ -0,0 +1,161 @@
+        return (3**f_ - 1)/2
+
+
+class SchurSubsetsNumber(Function):
```
I wonder what is schur subsets number?
And why is this needed symbolically if it is simply `log(2n+1, 3)`

mijo2 replied:
Well ```SchurSubsetsNumber``` gives us the upper bound of the number of sum free subsets ```k``` given the value ```n``` i.e. given the Schur Number ```S(k)``` what is the upper bound of k. I am little unsure whether it is needed symbolically or not, but the author of the previous PR #14493 used a function to return the value for computations required in ```schur_partition``` function.

sylee957 replied:
Unfortunately, I don't see any symbolic use of computations only for giving upper bounds. I'd expect it to be polished like `SchurNumber` or you should rather make this as a private function.

I'm thinking of an idea of having a function like `IntegersSumFreeSubsets` that can evaluate for known values, like how `SchurNumber` does now. And grouped with utility functions.

I think that the documentation is poor without a reference where `log(2n+1, 3)` appears. Though you may have to ask the original author or find the sources where it appears.

czgdp1807 replied:
> Unfortunately, I don't see any symbolic use of computations only for giving upper bounds.

Agreed.

mijo2 replied:
@sylee957 I would make it as a private function, but by private function you mean just using the formula and evaluating it right?

Also, I would ask the original author about the ```log(2n+1, 3)``` soon.

sylee957 replied:
For private function, you should prefix stuff with `_`

mijo2 committed:
combinatorics: Updated one of the functions to private function

czgdp1807 commented:
Tests needed for lines marked in red in [this page](https://codecov.io/gh/sympy/sympy/pull/18667/diff).

mijo2 committed:
combinatorics: Added new test cases for coverage of the code

mijo2 committed:
core: Removed SchurSubsetsNumber from test_args

czgdp1807 commented:
@sylee957 Can we merge this?
